codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameDecoder.java

```diff
@@ -212,69 +213,48 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
                     break;
                 }
 
-                final int idx = in.readerIndex();
+                final ByteBufChecksum checksum = this.checksum;
+                ByteBuf uncompressed = null;
 
-                ByteBuf uncompressed = ctx.alloc().heapBuffer(decompressedLength, decompressedLength);
-                final byte[] dest = uncompressed.array();
-                final int destOff = uncompressed.arrayOffset() + uncompressed.writerIndex();
-
-                boolean success = false;
                 try {
                     switch (blockType) {
-                    case BLOCK_TYPE_NON_COMPRESSED: {
-                        in.getBytes(idx, dest, destOff, decompressedLength);
-                        break;
-                    }
-                    case BLOCK_TYPE_COMPRESSED: {
-                        final byte[] src;
-                        final int srcOff;
-                        if (in.hasArray()) {
-                            src = in.array();
-                            srcOff = in.arrayOffset() + idx;
-                        } else {
-                            src = new byte[compressedLength];
-                            in.getBytes(idx, src);
-                            srcOff = 0;
+                        case BLOCK_TYPE_NON_COMPRESSED: {
+                            // Just pass through, we not not update the readerIndex yet as we do this outside of the
+                            // switch statement.
+                            uncompressed = in.retainedSlice(in.readerIndex(), decompressedLength);
+                            break;
                         }
+                        case BLOCK_TYPE_COMPRESSED: {
+                            uncompressed = checksum == null || checksum.isSupportingByteBuffer()
+                                    // We can allocate whatever buffer if we either not need to do checksum processing
+                                    // or if our ByteBufChecksum implementation supports ByteBuffer.
+                                    ? ctx.alloc().buffer(decompressedLength, decompressedLength)
+                                    : ctx.alloc().heapBuffer(decompressedLength, decompressedLength);
```

Comment:
@normanmaurer - is what we really mean that for jdk6 and jdk7 we may incur a copy operation in this case? for jdk8 this will be OK because we account for the [CR32](https://docs.oracle.com/javase/8/docs/api/java/util/zip/CRC32.html#update-java.nio.ByteBuffer-) and [Adler32](https://docs.oracle.com/javase/8/docs/api/java/util/zip/Adler32.html#update-java.nio.ByteBuffer-) class having the `update(ByteBuffer)` method, right? Something we can cleanup when supporting jdk8+ (maybe add an issue for cleanup later)?